### ISSUE TITLE
[Bug] TPS Server problems caused by ELN

### ISSUE THREAD

**Issue by codgamerMC** [Electrical-Age/ElectricalAge#817]
I recently encountered problems with ELN, my server reports a TPS of 3/20 and is now almost unplayable, my map is approx. 7500x3000 blocks it's quite big in a Flans mods car it will take you 5-10 min to get from tip to tip, I established an electrical grid all over the map, 5 Power Plants, it's a stable system and no tps problems until I decided to place streetlights all over the map in the 10 roads I built which are long roads, after the placement of over 2000 streetlights and using relays and signals to switch the on and off at night, I started having tps problems, and also I built a monitoring station for my power plant that uses aprox. 30 signal transmitters and it furthered the TPS into 3/20, I would like to know if this is a tick bug or something related in to that nature cause I really like the mod, I had to remove this station to get atleast 6-7 TPS but the main problem is the streetlights, are they supposed to be rendered all the time although player is not in that area? I also realized that streetlights stay on during day time on unloaded chunks that were loaded during the night and causing a jump in kW consumption sometimes leading to power plant security shutdown to protect the system and when restarting the lines, the instant kW and Amperage is too high to safely get the system voltage to 3.2 KV without it exploding, If I could know why the TPS issue happens and how to solve it I will appreciate it cause this map is my custom build on real life as close as possible, project is 3 years into the making and in the recent 4 months I have played with ELN and has changed it to the better!!

**Comment by lashtear** [comment]
Hm.  Several things:  Lighting changes are very heavy in minecraft, as they involve lighting recalculations; streetlighting of that sort then necessarily involves quite a lot of map churn and keeping everything chunkloaded makes for a very very large active set of chunks for the server.

Recommend:  Use a grid of poles, always energized (3.2kV on the ground, 12.8kV in the air).  Then supply individual lights via lamp supplies on dc-dc converters.  (200V LED bulbs most likely).  Do the switching for day/night via signals into the lamp supply; I recommend using a daylight sensor into a signal proc looking for e.g. <0.2.  With signal repeaters that can handle the whole grid on one sensor.  Anything doing switching on/off (lamps, processors, etc) needs to be chunkloaded.

Not doing chunkloading will cause it to continue with its last known state.

**Comment by codgamerMC** [comment]
Thanks for the comment, so few questions, I have a grid system like mentioned but every lets say 200 blocks I have a D.C. D.C. Converter with relay and daylight sensor with the lamp supply because each supply can only reach so far, so what you are saying is instead of using so many sensors just use one and use repeaters to get to the relays? Do I have to remove the lamp supplies or are they fine? Is he problem just the sensors? That is part of the problem the other part that remained unanswered is does the over 2000 streetlights with relays and sensors cause TPS problems cause I have been getting like 7-10 tps ever since, thanks in advanced

**Comment by lashtear** [comment]
I was primarily recommending use of a grid, with on/off switching handled by the supplies-- which you seem to do.  (Charging and discharging an entire grid is expected to be awkward because of the unfortunate grid capacitance in the current electrical model).

You can use separate sensors on each supply if you prefer; shouldn't make a big difference.  I do recommend switching via a wireless-signal to the supply rather than a relay to the supply but either should work.

Because of how Minecraft lighting updates work, the transition on/off or off/on will be somewhat lag-inducing.  That would show up as a temporary lag after each transition period.

If you're seeing persistent lag, that's different.  Can you describe whether that is server-side lag (e.g. handling all those tile entities) or client-side (e.g. rendering lag)?  "I don't know" is an okay answer, but we'll have to do more detailed testing.  Lowering your render distance may help check that though...

We've tested very large (continental) grids with diverse powerplants but I don't think we've performance tested large lighting arrays recently.  I'm sorta curious what the bottleneck there is.

**Comment by codgamerMC** [comment]
Well to let you know more about my grid system it's approx 7500 blocks wide by 3000 north, the grid supports more or less 2000 streetlights consumption of the grid is more than 60kW of power, after doing all this I noticed the server side lag in TPS and it wasn't client sided, my avg tick ranges from 130-230m/s which is bad, it's constant all the time and then I decided to create a monitoring station for the power plants and used over 30 transmitters and receivers, I changed the config to set the range of the transmitters to 10000 blocks so it could reach from one side of the map to the station without the use of repeaters and using that many transmitters made the lag worst into 3/20 TPS, what causes all this lag is beyond my knowledge and why was it caused, all I know is to my understanding my Processor can withstand this intensity but java isn't using it at all not even more than 2GB of RAM, which is all conserning, thanks in advanced!!

**Comment by jrddunbr** [comment]
Perhaps the best way to see what the problem is would be to replicate the problem. This would allow us to see how exactly the Minecraft server is behaving on local hardware, and what physical limitation we are hitting with the world you have.

Some useful information you could give us:
- Java command line, including flags, that you use to start your sever (or, in other words, how do you start Java)
- Platform:
  - Windows: Version Name
  - Linux: Output of ```uname -a```
- Version of the JRE you are using: ```java -version``` in a command prompt window or shell will do.
- Raw computer specs (ex, processor name, ram speed, size, and type, disk type) or if it's a hosted server, what the specs are that the hosting provider tells you.

My current suspicion is that the main thread of the Java program is hitting the top capacity of one processor core. Once that happens, the TPS begins to fall rather quickly, regardless of how many cores the system has - so your CPU usage may look low, but it may be already overloaded, and this is mostly undetectable without looking at how each core performs (and in general, CPU usage graphs are misleading). Having as much of the world loaded as you do may be causing the problem - if you have ~100,000 chunks loaded, that may be a limitation of the server hardware to do all of the ticks for each chunk loaded. Even 1/100'th of that being loaded (or 1,000 chunks) is a huge number of chunks loaded, and could definitely thrash a vanilla world on a decent server.

### Summary

If you can provide us system specs, the entire server directory (configs, world, mods) in an archive file (zip, 7z, tarball, etc), what command line you use to start it (or otherwise), and any other useful information, we can do some more complicated diagnostics locally using some debuggers and server tick monitors.

If you are not comfortable with giving us your world file, in replacement  you can give us a handful of screenshots with the critical components that drive your lighting systems, including but not limited to the distribution system, generation system, and other control systems you may have, so that we can re-create the problem. The other information such as the configs and mod-pack would still be useful.

I may ask more specific questions depending upon your platform, but this is all I can think of for now.

**Comment by codgamerMC** [comment]
I would love to provide the whole server directory only problem is it's about 2-3 GB files and you may or may not be aware, I live in Puerto Rico we been hit by a Hurricane on September and I have electricity but not internet so I rely on my phone ATT connection to connect my self, so I would spend my data plan on sending you that, what I can do is send many screenshots since they don't consume data and send that with all the information, i can comply with everything just that the world file is kind of big to send it, I appreciate all the help!! If you need more information after I send the information this afternoon you can reply and I will reply back, some useful information that you may now have that I can remember and later confirm is that I have Java 1.8073 version or similar, my of PC specs are i7-4790k 4ghz "8" cores 16GB RAM DDR3, 250 GB SSD with 3TB HDD, I will add to this info later on.

**Comment by codgamerMC** [comment]
I forgot to add that, my RAM speed is 1866 Mhz and Windows 7 ultimate 64bit

**Comment by jrddunbr** [comment]
Hmm, yeah don't waste your limited bandwidth. Mods can also be rather large, so instead just give us a complete listing of what is in the mods folder, and the Eln configuration file, and some screenshots.

As for CPU specs, [your CPU really dominates on single core performance - it's in the top 5 on this list.](https://www.cpubenchmark.net/singleThread.html) That is *probably* not the bottleneck. The rest of your rig looks pretty reasonable, so I think we've got the specs part down - it's probably not the computer's fault. I don't need any more platform info.

Likely, there are too many chunks being loaded, and having Eln talk across the entire world at once is causing a lot of chunk churn. Are you using chunk-loaders at all?

**Comment by codgamerMC** [comment]
Here you can download a zip file with all the relevant information I think might be necessary, just to point out, this server is for me a real life simulation I am trying to create, so it imitates Puerto Rico's Electrical Power Grid, you will find a Picture of Puerto Rico in dynmap form which is the shape of the actual Minecraft World and I used Paint to draw the grid I created, In yellow you will find the 3,2 KV grid which runs parallel thru all the roads and that's where the streetlight density can be found, almost all of the map is not loaded, chunk loaded areas are small, for example all power plants and some minor places here and there, there are approx. 5000 chunks loaded says OPIS (Also Included in the zip), in red the 12.8 KV Transmission lines, Green dots are distribution substations, in orange you will find the powerplants.  I also send pictures of the grid, pictures of the power plants, roads and streetlight setup, if you need more information let me know, it's a very big grid whose consumption is around 60kW.  http://www.mediafire.com/file/i4sczdv74n8edn3/ELN_Server_TPS_Problems.zip

**Comment by jrddunbr** [comment]
Run ```java -version```, and paste the output. You look to be using 32 bit Java if you are having problems using more than 2GB of ram. From what I see in Opis there, you may be just spending a ton of time GC'ing instead of doing game ticks... 13ms/t is fine for the world objects...

Also, can we have the Eln config you have? You didn't include it in the zip from what I see.

PS. Nice map :)

**Comment by codgamerMC** [comment]
Thanks!! Also thanks for all the help you given me I appreciate it a lot!! Here is the config file and one question, what is GC'ing... kind of ignorant on what that means.
# Configuration file

balancing {
    D:ElnToIndustrialCraftConversionRatio=0.3333333333333333
    D:ElnToOpenComputerConversionRatio=0.13333333333333333
    D:ElnToThermalExpansionConversionRatio=1.3333333333333333

    # Maximum horizontal distance from autominer that will be mined
    I:autominerRange=10
    D:batteryCapacityFactor=1.0
    I:fuelGeneratorPowerFactor=1
    D:fuelHeatFurnacePowerFactor=1.0

    # Factor to apply when converting real word heat values to Minecraft heat values (1mB = 1l).
    D:fuelHeatValueFactor=6.75E-5
    I:heatTurbinePowerFactor=1
    I:platesPerIngot=1
    I:solarPanelPowerFactor=1

    # Display more detailed WAILA info on some machines
    B:wailaEasyMode=false
    I:waterTurbinePowerFactor=1
    I:windTurbinePowerFactor=1
}


battery {
    # How many days it takes for a battery to decay half way
    I:batteryHalfLife=2
}


compatibility {
    B:ComputerProbeEnable=true
    B:ElnToOtherEnergyConverterEnable=true
}


debug {
    B:enable=false
}


dictionary {
    B:chips=true
    B:tungsten=false
}


entity {
    B:killMonstersAroundLamps=true
    I:killMonstersAroundLampsRange=9
    I:replicatorId=-1
    B:replicatorPop=true
    D:replicatorPopWhenThunderPerSecond=0.008333333333333333
}


fuelgenerator {
    I:tankCapacityInSecondsAtNominalPower=1200
}


gameplay {
    B:explosion=true
}


general {
    B:analyticsEnable=true
    S:playerUUID=b3a3dfcf-996f-4f3e-84e2-53c8a1e1eda9
    B:versionCheckEnable=true
}


lamp {
    D:carbonLifeInHours=6.0
    D:economicLifeInHours=64.0
    D:incandescentLifeInHours=16.0
    B:infiniteLedLife=false
    D:ledLifeInHours=512.0
}


mapgenerate {
    B:cinnabar=true
    B:copper=true
    B:forceOreRegen=false
    B:lead=true
    B:tungsten=true
}


modbus {
    B:enable=false
    I:port=1502
}


simulation {
    D:cableRsFactor=1.0
    I:electricalFrequency=20
    I:electricalInterSystemOverSampling=50
    I:thermalFrequency=400
}


wireless {
    I:txRange=10000
}


xrayscannerconfig {
    B:addOtherModOreToXRay=true
    B:canBeCrafted=true
    D:rangeInBloc=5.0
}

**Comment by codgamerMC** [comment]
Also on a side note, I tried the code: java -server -Xms4G -Xmx8G -jar custom_server.jar
It still didn't got beyond 2GB of RAM

**Comment by jrddunbr** [comment]
GC'ing means [Garbage Collect(ing)](https://en.wikipedia.org/wiki/Garbage_collection_(computer_science)). Java will occasionally go to collect the variables that are no longer in use and remove them from the ram allocation. If you run out of usable RAM within the JVM, it will race to clear as much RAM as possible, even if it's to the determent of the system running it. Eventually, it can just hang your stuff.

I still need your Java version, and whether it's 32 or 64 bit. Like I said before, ```java -version``` in a Windows command prompt will do the trick.

Glad to help!

**Comment by codgamerMC** [comment]
Java Version: 
![image](https://user-images.githubusercontent.com/10374236/33654083-6c2ee5da-da45-11e7-8d38-959956188af6.png)

**Comment by jrddunbr** [comment]
That's 64 bit Java alright. Now, what on Earth is your game doing if not playing the GC game... Perhaps a screenshot from each screen of Opis would help. The tick time spent on what looks to be nothing seems rather large. I'll have to experiment with that later this week.

**Comment by codgamerMC** [comment]
Do you have any suspicion at the moment of what can be that tick lag?
Something to note, I removed ELN from a backup I made and it instantly went up to 20TPS no lag at all, so I assume that the lag is caused by ELN in some way or form, how can you help me experiment on this issue?
 [Extra Info.zip](https://github.com/Electrical-Age/ElectricalAge/files/1536270/Extra.Info.zip)

**Comment by jrddunbr** [comment]
I suspect it must be something to do with the electrical simulation. Perhaps if you split your electrical network into smaller segments, it would behave better for the time being? Are all of the poles on the 3.2kv and 12.8kV networks connected together across the map? Perhaps just sticking to using local power generation to each quarter of the map would help the electrical simulation in some way, and isolate the networks otherwise, to prevent larger network calculations. Or, perhaps putting some of the DC/DC regulators in isolated mode would help.

The interesting thing I've noticed is that it appears that the time broken up for each tick is mostly spent on something that is not the worlds themselves, based on information in two of the screenshots on the first upload and confirmed by the second - meaning it's either the electrical simulation, or something else funky that we can't see. It's probably not directly related to the chunks, and if it is, it's too many chunks with Eln being loaded on it causing larger calculations for the network.

@Baughn - Do you have any ideas on what's up with the simulation? It's a black box to me.

**Comment by codgamerMC** [comment]
Yes all the lines are connected as shown in a grid color coded picture I sent you yesterday, I started noticing the problem when I was progressing with more and more streetlights, that meaning more lamp supplies, DC-DC converters, relays, amplifier and day light sensors, when I finished the whole map I noticed that TPS was around 7-10, are there any other programs other than OPIS that can see in detail what is going on, on a side note, this grid is really a large scale simulation.

**Comment by jrddunbr** [comment]
I was planning to put it into JVisual VM to watch usages. You can try that, too.

**Comment by codgamerMC** [comment]
I meant to ask you if you knew any programs on how to do that, I don't know how to. How do you watch usages?

**Comment by jrddunbr** [comment]
Well, if you run the program, you attach it to the currently running server process, and it will show what things are taking what amount of time ("usages" - referring to using a profiler for CPU time). If you see something out of the ordinary while the CPU profiler is running, then you analyze it further. Generally, whatever is lowering your TPS will be repeating a lot and/or taking a lot of time.

Note that there is a lot to watch, there are network threads, tick threads, and other threads related to the GUI, since you're using it in headed mode on Windows (versus running headless mode Java like I usually do in Linux; in that platform, you wouldn't encounter those threads running.) Some mods may also be breaking out into other threads.

Perhaps something like [this](https://www.youtube.com/watch?v=Qb1x0uFGJbY) would be helpful.

**Comment by cm0x4D** [comment]
@codgamerMC You can try to set some DC/DC converters into "isolation" mode, so the huge simulated grid will be partitioned in smaller subnets and the Electrical Age simulation should theoretically get calculated more efficiently. The downside of this is that the simulation can get a bit less reactive as if there were big capacitors placed on both sides of the DC/DC converter.

**Comment by codgamerMC** [comment]
Currently I am busy on some work I have to do as of today I can't do more testing, hopefully tomorrow I can do some and I will get back to you guys tomorrow if I can and thanks for all the help, I really appreciate it

**Comment by Baughn** [comment]
If he uses utility poles, those are always isolated. (At the transformer pole / ground connection, that is.)

**Comment by codgamerMC** [comment]
At the root of each lamp supply there is the poles then a transformer that is in not isolated mode, relay, lamp supply, day light sensors with amplifiers, this setup repeats I would estimate for over 100 times, since I did this across the map I have been getting this lag, I removed the mod on a backup I made to test and it jumped back to 20 TPS so I assume there is something in the mod that is causing this mysterious decay in TPS, I also noticed that if you use more than 30 signal transmitters in the map over some distance, that will down the server on 6 more TPS, as to why, I don't know, any ideas @Baughn?

**Comment by codgamerMC** [comment]
Also thanks for the huge help to everybody that has helped in this thread, I love this mod!!

**Comment by bloxgate** [comment]
>I also noticed that if you use more than 30 signal transmitters in the map over some distance, that will down the server on 6 more TPS, as to why, I don't know, any ideas

That's really odd. I remember playing on servers with way more than 30 signal transmitters with no issues. I've noticed you have the transmitter range set to 10 km, perhaps the transmitters are very inefficient with such a large range? Could you try using a significantly smaller range with repeaters, or transmitting your control signals via signal cable?

**Comment by codgamerMC** [comment]
I have no problem but the thing is that I built like a command center to receive this signals, but the default 32 blocks wouldn't help to transmit the signal over 6 km since my map is like 10 km wide I set it to that, but I had the TPS problems, way before that, anyway to test Tick time other than opis? Thanks for the help

**Comment by bloxgate** [comment]
You can transmit signals more than 32 blocks using the default range, you just need to put signal repeaters down near the edge of the range.

**Comment by Baughn** [comment]
The signal blocks raytrace to each other in order to figure out if they're in range or not, and they're not clever about caching data from unloaded chunks. I can absolutely believe that setting an excessively large range would cause lag.

**Comment by codgamerMC** [comment]
But aside that, that wasn't the real problem of the TPS, the problems happened way before I tried adding those, any idea on what it might be? Still thinking it has something to do with the streetlights, in my opinion

**Comment by Baughn** [comment]
It's hard to say for sure without profiling the server. Which is to say--

Can you run [Warmroast](https://github.com/sk89q/WarmRoast) against it, and check where the CPU time goes? Chances are it's fairly obvious in the profile, but go ahead and take a screenshot once you have it working.

**Comment by codgamerMC** [comment]
In what profile you say it's obvious in the Opis one or the warmroast? Later on today I will run the test

**Comment by Baughn** [comment]
Warmroast. I won't guarantee it'll be obvious, but it'll at least give us somewhere to start.

Opis profiling is useful, but it overlooks a great deal. Warmroast will pick up the entire runtime.

**Comment by codgamerMC** [comment]
I have been trying for 2 hours to start that damn program but it won't start nor are there any tutorials to aid the installation, could you tell me how to run this damn program, I want to know already what is causing the lag, Thanks for the help by the way

**Comment by codgamerMC** [comment]
Some Files from VisualVM, until I can get running the other program this is the best I got for now, any leads? I switched the world into single player to see, on the second post I will post from the server side
![cpu profiling](https://user-images.githubusercontent.com/10374236/33800173-705296ce-dd11-11e7-8a76-68997a8a344b.PNG)
![cpu profiling2](https://user-images.githubusercontent.com/10374236/33800174-7065dc3e-dd11-11e7-8019-96a5876bdb47.PNG)
![cpu profiling3](https://user-images.githubusercontent.com/10374236/33800175-7078e2a2-dd11-11e7-8b48-a0b29ef86e6a.PNG)
![cpu profiling4](https://user-images.githubusercontent.com/10374236/33800176-708d7dca-dd11-11e7-8253-4fba4b516b8b.PNG)
![cpu profiling5](https://user-images.githubusercontent.com/10374236/33800177-70a2632a-dd11-11e7-865b-071b665d8832.PNG)

**Comment by codgamerMC** [comment]
World on the Server Side
![serverside](https://user-images.githubusercontent.com/10374236/33800234-4af61044-dd12-11e7-9a48-d56e15e0e130.PNG)
![serverside2](https://user-images.githubusercontent.com/10374236/33800235-4b0c5ce6-dd12-11e7-8694-179139ff4801.PNG)
![serverside3](https://user-images.githubusercontent.com/10374236/33800236-4b1ced22-dd12-11e7-94b7-9ceed3c8fa1a.PNG)
![serverside4](https://user-images.githubusercontent.com/10374236/33800237-4b2e3da2-dd12-11e7-8c3c-4d8ebb210998.PNG)

**Comment by jrddunbr** [comment]
The first server screenshot is a godsend. It shows the entire problem.

So, it appears that wireless signal transmitters are using a lot of your CPU time. Specifically:

The wireless transmitters requires the world to be loaded between them, and does a block transparency ray-trace. This loads all of the chunks which have a wireless device in them, and every chunk between them, every tick. Ouch. 

Perhaps that calculation should be done when you place them, to save TPS; or make the distance cost calculation optional with a config setting, ex, perhaps if the config is ```-1```km, it just disables the distance calculation, saving the TPS, since it doesn't calculate the distance traveled. We may want to put a comment in the config mentioning that distances over 64m is probably bad for your TPS.

TL;DR, having the wireless transmitter distance set to 10km is bad.

PS. The solve() function is also taking a lot of time, but you have a big grid, so it makes sense. That's probably not going to negatively affect your TPS as much

[This raytrace is what created the problem](https://github.com/Electrical-Age/ElectricalAge/blob/6f6d086fd98900bf8c400497d39bb3dae88b1d07/src/main/java/mods/eln/sixnode/wirelesssignal/WirelessUtils.java#L158)

**Comment by codgamerMC** [comment]
So what I should do is put the transmitter distance to -1, but how does that explain the lag I started experiencing before having any transmitters, for what I know I don't have any or maybe 1 or 2(in a close distance like 100 blocks), how does that cause the lag?

**Comment by codgamerMC** [comment]
Update: Setting the config to -1000 upped the TPS to 10 but still not great

**Comment by codgamerMC** [comment]
One question: is there anyway for you to test for any problems by joining the server? If so I could find a way to establish a connection

**Comment by jrddunbr** [comment]
Heh, whops. Should have been more specific. I was suggesting it as a potential fix I could write. Sorry for the confusion, such a setting does not exist in the code.

We're figuring out a long-term solution now, looks like by caching the ray-trace. That will solve the problem with the wireless transmitters.

For the time being, probably the best solution is to set the wireless range to the normal 32 blocks, and use a signal wire to transmit the signal until it's fixed.

Besides that, I'd have to look closer to see what's hurting the TPS besides that, but in the meantime, change the range and see if that helps it at all. If not, take some new screenshots of the first server side graph, with the range set to ```32```.

**Comment by jrddunbr** [comment]
I don't think there's any great way for me to debug it by joining the server. Not effectively, at least.

**Comment by codgamerMC** [comment]
Alright, I will do that just now so hold on, but hey atleast you can see whats going on and see the type of grid I have setup

**Comment by codgamerMC** [comment]
PS: Setting it to 32 doesn't help much the TPS but I will do another run of VisualVM

**Comment by codgamerMC** [comment]
You sir just helped me get this problem trough, the TPS Throttled up to 14-17/20, still not 20 but hey much better, looks like the signal transmitters cause big problems, as to why 5 ticks deficient I will send a picture now

**Comment by codgamerMC** [comment]
New Sampler 14-17 TPS
![image](https://user-images.githubusercontent.com/10374236/33800621-e6008b76-dd19-11e7-84c2-f76b00f37b1f.png)
![image](https://user-images.githubusercontent.com/10374236/33800623-ef2e9da0-dd19-11e7-92d3-d26fd798780d.png)
![image](https://user-images.githubusercontent.com/10374236/33800627-f6680a98-dd19-11e7-8e7e-c853a5c571ad.png)
![image](https://user-images.githubusercontent.com/10374236/33800638-4228fcd0-dd1a-11e7-8065-c69053ca41ce.png)

**Comment by jrddunbr** [comment]
Whoa, that's fantastic! Glad I could help.

PS. I only need the top graph of the three you just sent, and I need it separate from the client to get a good idea of what's going on (ie, run the server standalone)

**Comment by codgamerMC** [comment]
Hold on right there, it will take me a few min I will do it right now

**Comment by codgamerMC** [comment]
PS: Part of the reason why I switched to single player was because I am having problems allocating more than 2 GB of RAM to the server client, even though it is a 64 bit, which reduced TPS by 2-3
Code: c:\progra~1\Java\jre1.8.0_73\bin\java.exe -server -Xms4G -Xmx8G -jar custom_server.jar

**Comment by codgamerMC** [comment]
New Pictures, any thoughts?
![image](https://user-images.githubusercontent.com/10374236/33800685-87e06bc2-dd1b-11e7-8983-a9d778c25abc.png)
![image](https://user-images.githubusercontent.com/10374236/33800688-915b3f7e-dd1b-11e7-9467-f06596fe9cfb.png)

**Comment by codgamerMC** [comment]
I am starting to think that probably I have tested this mod to it's extremes by going beyond what anyone has done and for all I have done, this isn't such a big deal, one thing to note is that the transmitters cause lag beyond configured and you guys for next versions should at least design a way to transmit signal farther than 32 blocks because that's what I wanted to do to build a monitoring station to monitor all my Power Plants and in an attempt to cross 7 km of world I changed the config and that happened so a little suggestion would be to work on that because I really want to be able to redo that station and get all data, I thought of doing utility poles but the cables are not compatible so... that happened, but thanks for all the help I really appreciate it!!

**Comment by bloxgate** [comment]
If you really want to transmit a signal wirelessly over that large of a distance rather than just doing a massive cabling run, the computer probe would probably be helpful. It'd be a bit expensive, but something like an OpenComputers entangled card would allow for long distance communication between two devices.

As for the rest of the lag, from the latest profiler screenshot it seems like your server is spending most of the tick-time processing Electrical Age's simulation, **but** it looks like there's a lot of IO going on for the world. Are you sure that you don't have something loading and unloading a lot of chunks still?

As for the RAM problem, do you have any global java options set that might be limiting it to something other than what you're giving it in its launch arguments?

**Comment by codgamerMC** [comment]
For the IO, can't remember anything, I only have chunk loaders on the power plants, I might need to do some research, the java arguments, nothing seems to be limited because Minecraft Launcher can start with more than 2GB, so I have no idea there.

**Comment by codgamerMC** [comment]
How would I go about using the ELN computer probe?

**Comment by Baughn** [comment]
For diagnosing chunk-churn, one useful tool is to run the save-off command in your server. If that helps TPS, then something is loading and unloading chunks a lot. It's unlikely to be Eln, though--Eln blocks don't need to be chunkloaded, but with a few exceptions they also won't trigger chunkloading.

Be careful! Turning off saving prevents the server from unloading chunks, which in effect means you've created a memory leak. Additionally, it will *prevent the server from saving chunks during shutdown*. This can corrupt your world; always turn saving back on before stopping the server.

If it does help, then one bandaid is to keep saving off most of the time but turn it on for half a minute every ten minutes or so. You'll need the save-all command in that case.

(Careful again! Save-all does nothing if saving is off.)

**Comment by AMIDIBOSS** [comment]
Does "the computer probe" work with CC?

**Comment by jrddunbr** [comment]
Yes.

**Comment by codgamerMC** [comment]
How can I use it?

**Comment by jrddunbr** [comment]
I would write you some lua code, but I have my semester's final exams this week (thus the sluggish response on all of my communications). Perhaps later I can write something.

Effectively, if you use the [tunnel card](http://ocdoc.cil.li/item:linked_card) from opencomputers, you can transmit data over any distance without introducing lag. From there, a simple lua program just would pipe data from the ["ElnProbe"](https://wiki.electrical-age.net/index.php?title=Eln_Computer_Probe) object in lua (using mostly the ```signal*``` methods) to the tunnel, and from the tunnel card in the other computer back to the ElnProbe, and you can do both input and output on those probes. From there, you just connect the wires where you need, and tell the computers to always send available data to the other probe. **Note** - The tunnel card has a limited implementation of the network api - not all programs made for the OC/CC network will work for the tunnel card, and vica versa.

If you need some inspiration on how to use OpenComputers with ELN, check out my github repo called [oc-eln](https://github.com/jrddunbr/oc-eln). From there, just follow the API documentation and Lua documentation on the OpenComputers website (see above). Potentailly, my friend Grissess (that's his github username) also has some useful code, such as a [hardware proxy daemon over an opencomputers network](https://github.com/Grissess/cc-progs/blob/master/daemons/hwproxyd). None of thse are plug and play, but bits and pieces could be used rather effectively here, and it's what I would work from to create the code necessary.

You *can* use ComputerCraft for this as well, and the ELN probe will behave mostly the same from my understanding, but they don't have the tunnel cards, so you would have to use the standard network, which means you need to run a CC cable (or use AE2 computercraft tunnels, which *still* require a AE2 wire to run the distance)

Protip: ```component.get()``` and ```component.proxy()``` are super useful.

**Comment by jrddunbr** [comment]
Heya,

I wrote some rather jank (but workable) Lua code (it's very not production level)

Probably someone who is more experienced in Lua could make something more polished, but this at least *works now*... 

[https://github.com/jrddunbr/oc-eln/tree/master/elnProbeLink](https://github.com/jrddunbr/oc-eln/tree/master/elnProbeLink)

If you set up two computers with internet cards (so that you can download the file from the internet - use the GitHub Raw download blob thing), a linked card in each one, and then a computer probe connected to each computer, you can transmit Eln signals from one computer to another. Start the server program on the server and the client program on the client, and it will update the probe (supposedly) once per tick. (or every 0.05 seconds is how I programmed it - that remains to be seen tho)

Have fun!

Edit: [inserts photo] By the way, it appears to be a bit slow compared to the Eln signals, but it does still get the values right, and it's mostly alright.

![2017-12-12_16 49 59](https://user-images.githubusercontent.com/14130505/33910518-00d49812-df5d-11e7-9793-413651603fdd.png)

**Comment by codgamerMC** [comment]
Thanks for all the help, I am currently really busy on a lot of crap like you are(understandable), I invite you to join my server sometime after we get done with work so you can walk around and see it and help me setup some of this, that is if you want to come and see the server, aside from that thanks for all the help really, exited to try this soon!!

**Comment by jrddunbr** [comment]
Sure, not a problem, love to help.

I probably can't commit to doing anything, but perhaps next week I could poke around on your server and see what things look like.

If you want to share the server information privately, just shoot me an email - it's my GitHub username at gmail, so ya.

**Comment by codgamerMC** [comment]
Yeah I know, I can't this week either, like I said busy also, but I am hoping I am in luck, my local cable company installed new lines and connected mine already, now we are waiting for Power Authority to connect the electricity to the boxes and then a recovery and then I have internet back, so hopefully it gets back and we can play then and setup some things, thanks for the help man!!

**Comment by codgamerMC** [comment]
If you want to check out my server I will be free to do so later this week

**Comment by jrddunbr** [comment]
Sounds like a plan. I'm working on a PR for the fix to make it perform better if one does set that range, but it's taking a while because I'm sorta new to Minecraft modding.

**Comment by codgamerMC** [comment]
I am free today, would you mind bumping into the server for a while if so I will send you an email with the details.

**Comment by jrddunbr** [comment]
Shoot me an email: ```jrddunbr``` aht ```gmail.com```

**Comment by codgamerMC** [comment]
I sent you an email make sure you have all the mods necessary I will send you a second email with the mods

**Comment by jrddunbr** [comment]
I looked through the code a bit, this also affects the performance of wireless lighting systems I believe. If you set the wireless signal distance to 10k, it will also let you use a light fixture's power channel 10k blocks away. This has some interesting implications. lol

So yeah, that also can be caused by this variable being changed. Hopefully, I can finish the PR by the end of the week, if I find time to work on it.

The interesting question - what happens when you have hundreds of lights on the same channel, and how much of a mess does one make, based on voltages and stuff. If you have two lighting systems on one channel, and one is 50v and one is 60v, do all of the lights swap back and forth, pick the highest voltage, average, etc, etc. Ray-tracing everywhere! :P
